clouddriver-ecs/src/main/java/com/netflix/spinnaker/clouddriver/ecs/provider/agent/IamRoleCachingAgent.java

```diff
@@ -0,0 +1,172 @@
+import com.amazonaws.services.identitymanagement.model.Role;
+import com.netflix.spinnaker.cats.agent.AgentDataType;
+import com.netflix.spinnaker.cats.agent.CacheResult;
+import com.netflix.spinnaker.cats.agent.CachingAgent;
+import com.netflix.spinnaker.cats.agent.DefaultCacheResult;
+import com.netflix.spinnaker.cats.cache.CacheData;
+import com.netflix.spinnaker.cats.cache.DefaultCacheData;
+import com.netflix.spinnaker.cats.provider.ProviderCache;
+import com.netflix.spinnaker.clouddriver.aws.security.AmazonClientProvider;
+import com.netflix.spinnaker.clouddriver.ecs.cache.Keys;
+import com.netflix.spinnaker.clouddriver.ecs.cache.model.IamRole;
+import com.netflix.spinnaker.clouddriver.ecs.provider.EcsProvider;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static com.netflix.spinnaker.cats.agent.AgentDataType.Authority.AUTHORITATIVE;
+import static com.netflix.spinnaker.clouddriver.ecs.cache.Keys.Namespace.IAM_ROLE;
+
+public class IamRoleCachingAgent implements CachingAgent {
+
+  static final Collection<AgentDataType> types = Collections.unmodifiableCollection(Arrays.asList(
+    AUTHORITATIVE.forType(IAM_ROLE.toString())
+  ));
+
+  public static final String DEFAULT_IAM_REGION = "us-east-1";
+
+  private final Logger log = LoggerFactory.getLogger(getClass());
+  private AmazonClientProvider amazonClientProvider;
+  private AWSCredentialsProvider awsCredentialsProvider;
+  private String accountName;
+  private IamPolicyReader iamPolicyReader;
+
+
+  public IamRoleCachingAgent(String accountName,
+                             AmazonClientProvider amazonClientProvider,
+                             AWSCredentialsProvider awsCredentialsProvider,
+                             IamPolicyReader iamPolicyReader) {
+    this.accountName = accountName;
+    this.amazonClientProvider = amazonClientProvider;
+    this.awsCredentialsProvider = awsCredentialsProvider;
+    this.iamPolicyReader = iamPolicyReader;
+  }
+
+
+  @Override
+  public CacheResult loadData(ProviderCache providerCache) {
+    AmazonIdentityManagement iam = amazonClientProvider.getIam(accountName, awsCredentialsProvider, DEFAULT_IAM_REGION);
+
+    Set<IamRole> cacheableRoles = fetchIamRoles(iam);
+    Map<String, Collection<CacheData>> newDataMap = generateFreshData(cacheableRoles);
+    Collection<CacheData> newData = newDataMap.get(IAM_ROLE.toString());
+
+    Set<String> oldKeys = providerCache.getAll(IAM_ROLE.toString()).stream()
+      .map(cache -> cache.getId()).collect(Collectors.toSet());
+    Map<String, Collection<String>> evictionsByKey = computeEvictableData(newData, oldKeys);
+
+    DefaultCacheResult cacheResult = new DefaultCacheResult(newDataMap, evictionsByKey);
+    return cacheResult;
```

Comment:
Please add logging.

Ei. How many cached, how many evicted.

See other agents for examples.